### PR TITLE
SP4 first commit

### PR DESCRIPTION
**Pull request by skrutt**
Added support for SP4 type plugs. Tested with SP4L-EU type, seems to share alot with US and AU types, as EU plug was not present in IHC app but i could control my EU plug anyway.
Added extra support functions check_power/set_power to comply with sp1,sp2,sp3 type controls.

Also updated docs with some(small) info

### CODE REVIEW

**Comment by skrutt** [comment]
Only tested on one device, but hopefully solves #299

**Comment by nihilus** [comment]
Work for me although I had to add 'state = json.loads(payload[0x0c:0x0c+js_len].decode("utf-8"))', you see the decoding part there? :+1:

**Comment by skrutt** [comment]
@nihilus Cool! im guessing you use python2, i only tried it with python3. If you are using python 2, maybe i should add that to the PR for compability?

**Comment by nihilus** [comment]
I do use python 3.5 for the record.

**Comment by skrutt** [comment]
Hm. Now that you say that i think that i saw somewhere that after python 3.5, decode is implicit? So it's probably a good idea either case.

**Comment by andwaal** [comment]
Amazying work, looking forward to this getting merged 🖒

**Comment by ristohelkio** [comment]
Got this working with a Clas Ohlson 'WiFi Smart Plug'. Thumbs up for getting this merged.

File: broadlink/__init__.py [hunk 1]
```diff
@@ -284,6 +287,88 @@ def send_packet(self, command, payload):
...
+        if indicator is not None:
+            data['indicator'] = int(bool(indicator))
+        if ntlbrightness is not None:
+            data['ntlbrightness'] = ntlbrightness
+        if maxworktime is not None:
+            data['maxworktime'] = maxworktime
+
+        js = json.dumps(data).encode('utf8')
+        packet = self._encode(2, js)
+        response = self.send_packet(0x6a, packet)
```
Comment (felipediel) [inline broadlink/__init__.py:317]: Use this function to check for errors in the response:
```python3
check_error(response[0x22:0x24])
```

File: broadlink/__init__.py [hunk 2]
```diff
@@ -284,6 +287,88 @@ def send_packet(self, command, payload):
...
+        for c in packet:
+            checksum = (checksum + c) & 0xffff
+        packet[0x04] = checksum & 0xff
+        packet[0x05] = checksum >> 8
+        return packet
+
+    def _decode(self, response):
+        err = response[0x22] | (response[0x23] << 8)
+        if err != 0:
+            return None
```
Comment (felipediel) [inline broadlink/__init__.py:365]: Remove this.

**Comment by felipediel** [comment]
I also think it is important to add the `.decode("utf-8")` mentioned above, so as not to break Python 2 support.

**Comment by yusijs** [comment]
Ooh this is awesome. @skrutt will you be able to resolve the changes @felipediel requested? I have a few of these plugs as well and would love being able to integrate them with my (new) homeassistant setup :D 

Or would you prefer someone else pick it up if you don't have the time?

**Comment by Tamarinen** [comment]
As I have one of those switches, I've taken the liberty of porting @skrutt 's code to the new structure in https://github.com/mjg59/python-broadlink/pull/414 . I believe I've incorporated all changes suggested here in that.

**Comment by felipediel** [comment]
@skrutt We adapted and merged [here](https://github.com/mjg59/python-broadlink/pull/429). Thank you very much for this!

This PR can be closed.

**Comment by felipediel** [comment]
We resolved conflicts and merged [here](https://github.com/mjg59/python-broadlink/pull/429). Thank you very much @skrutt!

**Pull request closed without merging**